x/mongo/driver/topology/server.go

```diff
@@ -294,27 +292,64 @@ func (s *Server) RequestImmediateCheck() {
 	}
 }
 
+// ProcessHandshakeError handles connection errors during the handshake.
+func (s *Server) ProcessHandshakeError(err error, conn driver.Connection) {
+	// ignore nil or stale error
+	if err == nil || conn.Stale() {
+		return
+	}
+
+	wrappedConnErr := unwrapConnectionError(err)
+	if wrappedConnErr == nil {
+		return
+	}
+
+	// Since the only kind of ConnectionError we receive from pool.Get will be an initialization
+	// error, we should set the description.Server appropriately.
+	s.updateDescription(description.NewServerFromError(s.address, wrappedConnErr, s.Description().TopologyVersion))
+	s.pool.clear()
+}
+
 // ProcessError handles SDAM error handling and implements driver.ErrorProcessor.
-func (s *Server) ProcessError(err error) {
-	desc := s.Description()
+func (s *Server) ProcessError(err error, conn driver.Connection) {
+	s.processErrorLock.Lock()
+	defer s.processErrorLock.Unlock()
+
+	// ignore nil or stale error
+	if err == nil || conn.Stale() {
```

Comment:
Let's move the `err == nil` check out of the lock. It's a small thing, but the majority of operations should not get an error, and we want to keep the happy path lock-free when possible.